File: deno/src/wx/pay_transactions_jsapi/pay_transactions_jsapi.dao.ts
import {
  log,
} from "/lib/context.ts";

import {
  getAuthModel,
} from "/lib/auth/auth.dao.ts";

import {
  findByIdWxUsr,
  validateOptionWxUsr,
} from "/gen/wx/wx_usr/wx_usr.dao.ts";

// wx_wxo_usr
import {
  findByIdWxoUsr,
  validateOptionWxoUsr,
} from "/gen/wx/wxo_usr/wxo_usr.dao.ts";

import {
  findByIdTenant,
  validateOptionTenant,
} from "/gen/base/tenant/tenant.dao.ts";

import {
  getWxPayModel,
} from "/src/wx/wx_pay/wx_pay.dao.ts";

import {
  findOneWxPay,
  validateOptionWxPay,
} from "/gen/wx/wx_pay/wx_pay.dao.ts";

import {
  createPayTransactionsJsapi,
} from "/gen/wx/pay_transactions_jsapi/pay_transactions_jsapi.dao.ts";

import {
  findOneDomain,
  validateOptionDomain,
} from "/gen/base/domain/domain.dao.ts";

import type {
  RequestPaymentOptions,
} from "/gen/types.ts";

import {
  PayTransactionsJsapiTradeState,
  PayTransactionsJsapiCurrency,
} from "/gen/types.ts";

import WxPay from "wechatpay-node-v3";

import {
  Buffer,
} from "node:buffer";

/** 订单金额信息 */
interface Iamount {
  /** 订单总金额，单位为分 */
  total: number;
  /** 货币类型 */
  currency?: string;
}

/** 支付者信息 */
interface Ipayer {
  openid: string;
}

/** 场景信息 */
interface IsceneInfoNative {
  /** 服务器IP地址 */
  payer_client_ip: string;
  /** 商户端设备号 */
  device_id?: string;
  store_info?: IstoreInfo;
}

// 商户门店信息
interface IstoreInfo {
  id: string;
  name?: string;
  area_code?: string;
  address?: string;
}

// 单品列表信息
interface IgoodsDetail {
  merchant_goods_id: string;
  wechatpay_goods_id?: string;
  goods_name?: string;
  quantity: number;
  unit_price: number;
}

// 优惠功能
interface Idetail {
  cost_price?: number;
  invoice_id?: string;
  goods_detail?: IgoodsDetail[];
}

interface Ijsapi {
  description: string;
  out_trade_no: string;
  time_expire?: string;
  attach?: string;
  notify_url: string;
  goods_tag?: string;
  amount: Iamount;
  payer: Ipayer;
  detail?: Idetail;
  scene_info?: IsceneInfoNative;
}

interface Ipay {
  appid: string; //  直连商户申请的公众号或移动应用appid。
  mchid: string; // 商户号
  serial_no?: string; // 证书序列号
  publicKey: string; // 公钥
  privateKey: string; // 密钥
  authType?: string; // 认证类型，目前为WECHATPAY2-SHA256-RSA2048
  userAgent?: string;
  key?: string;
}

/**
 * 统一下单
 */
export async function transactions_jsapi(
  appid: string,
  params0: Partial<Ijsapi>,
  attachInfo?: {
    attach2?: string;
  },
): Promise<RequestPaymentOptions> {
  const {
    obj,
    params,
  } = await getJsapiObj(appid);
  Object.assign(params, params0);
  log(`transactions_jsapi.params: ${ JSON.stringify(params) } `);
  const wxPay = new WxPay({
    appid: obj.appid,
    mchid: obj.mchid,
    publicKey: Buffer.from(new TextEncoder().encode(obj.publicKey)),
    privateKey: Buffer.from(new TextEncoder().encode(obj.privateKey)),
    key: obj.key,
  });
  const res = await wxPay.transactions_jsapi(params);
  log(`transactions_jsapi.result: ${ JSON.stringify(res) }`);
  const result = res.data as unknown as RequestPaymentOptions;
  /*
  {
    "status": 200,
    "appId": "wx28e52faff22e8837",
    "timeStamp": "1675853534",
    "nonceStr": "nhyc2alf7qt",
    "package": "prepay_id=wx08185215073441b9904f2b8ef267bf0000",
    "signType": "RSA",
    "paySign": "D7h2C0j2hWGdC3RG8lkBEv/b5GmEla9nzoJtRhSjREz5pCYbj1FEuibDTdyKQCGc9RjgU1Tf1YbZg9/9FPHns5FsZVghx9FqkqmwVs+xeT1jQs6V9PT50mb49kQjtw74nFXGvqI37APcLRIzTy2bbWk6eVnX0eB6oWzzBkQfKF31OXy7REZoTKRhQl/A/OT5DcjVIoLRR+DVYhGJ3Fn9c3YuI6/wQ+owENkgvpKibPzBgD9IMgN/UNzWb+9H7Ae/a8T594TIRsfAENbucqMHHLng7xvOon88LAzUZ5DuiyMTKQJyBoQ0x77eRPEuXMVkv9oXm9TiWUrFFxaoskDx3w=="
  }
  */
  type PayMsg = {
    code: string;
    message: string;
    status: number;
  };
  if ((result as unknown as PayMsg).status !== 200) {
    const data = result as unknown as PayMsg;
    throw `${ data.code }: ${ data.message }`;
  }
  result.orderInfo = obj.appid;
  const authModel = await getAuthModel();
  const wx_usr_id = authModel?.wx_usr_id;
  
  if (!wx_usr_id) {
    throw new Error("transactions_jsapi.wx_usr_id is null");
  }
  
  const wx_usr_model = await validateOptionWxUsr(
    await findByIdWxUsr(wx_usr_id),
  );
  const tenant_id = wx_usr_model.tenant_id;
  
  if (!params.notify_url) {
    throw "notify_url 未设置";
  }
  
  await createPayTransactionsJsapi({
    appid: obj.appid,
    mchid: obj.mchid,
    description: params.description,
    out_trade_no: params.out_trade_no,
    transaction_id: "",
    trade_state: PayTransactionsJsapiTradeState.Notpay,
    trade_state_desc: "未支付",
    time_expire: params.time_expire,
    attach: params.attach,
    attach2: attachInfo?.attach2,
    notify_url: params.notify_url,
    // receipt: ,
    // profit_sharing,
    total_fee: params.amount.total,
    currency: params.amount.currency as unknown as PayTransactionsJsapiCurrency,
    openid: params.payer.openid,
    prepay_id: result.package,
    tenant_id,
  });
  return result;
}

export async function getJsapiObj(
  appid: string,
) {
  
  const wx_pay_model = await validateOptionWxPay(
    await findOneWxPay({
      appid,
    }),
  );
  await getWxPayModel(wx_pay_model);
  let notify_url = wx_pay_model.notify_url;
  const mchid = wx_pay_model.mchid;
  const public_key = wx_pay_model.public_key;
  const private_key = wx_pay_model.private_key;
  const v3_key = wx_pay_model.v3_key;
  const payer_client_ip = wx_pay_model.payer_client_ip;
  
  const auth_model = await getAuthModel();
  const wx_usr_id = auth_model?.wx_usr_id;
  const wxo_usr_id = auth_model?.wxo_usr_id;
  
  let openid: string | undefined = undefined;
  let tenant_id: TenantId | undefined = undefined;
  
  if (wx_usr_id) {
    const wx_usr_model = await validateOptionWxUsr(
      await findByIdWxUsr(wx_usr_id),
    );
    openid = wx_usr_model.openid;
    tenant_id = wx_usr_model.tenant_id;
  } else if (wxo_usr_id) {
    const wxo_usr_model = await validateOptionWxoUsr(
      await findByIdWxoUsr(wxo_usr_id),
    );
    openid = wxo_usr_model.openid;
    tenant_id = wxo_usr_model.tenant_id;
  }
  if (!openid) {
    throw "openid 未设置";
  }
  if (!tenant_id) {
    throw "tenant_id 未设置";
  }
  
  const tenantModel = await validateOptionTenant(
    await findByIdTenant(tenant_id),
  );
  if (tenantModel.is_enabled !== 1) {
    throw `租户 ${ tenantModel.lbl } 未启用`;
  }
  const domain_ids = tenantModel.domain_ids;
  if (!domain_ids || domain_ids.length === 0) {
    throw `租户 ${ tenantModel.lbl } 未设置默认域名`;
  }
  const domainModel = await validateOptionDomain(
    await findOneDomain({
      ids: domain_ids,
      is_default: [ 1 ],
      is_enabled: [ 1 ],
    }),
  );
  const domain_lbl = domainModel.lbl;
  const protocol = domainModel.protocol || "https";
  
  notify_url = `${ protocol }://${ domain_lbl }${ notify_url }`;
  
  const out_trade_no = crypto.randomUUID().replaceAll("-", "");
  const obj = {
    appid,
    mchid,
    publicKey: public_key,
    privateKey: private_key,
    key: v3_key,
  };
  return {
    obj,
    params: {
      description: "商品描述", // 商品描述 会被替换
      amount: {
        total: 1, // 订单总金额，单位为分 会被替换
      },
      out_trade_no,
      notify_url,
      payer: {
        openid,
      },
      scene_info: {
        payer_client_ip,
      },
    } as Ijsapi,
  };
}

File: rust/src/src/wx/pay_transactions_jsapi/pay_transactions_jsapi_dao.rs
use color_eyre::eyre::{Result, eyre};

use crate::common::context::{
  Options,
  get_short_uuid,
  get_auth_model_err,
};
use crate::common::wx_pay::{Amount, Jsapi, Payer, WxPayData, WxPay, SceneInfo};

use super::pay_transactions_jsapi_model::RequestPaymentOptions;

// wx_pay
use crate::r#gen::wx::wx_pay::wx_pay_dao::{
  find_one_wx_pay,
  validate_option_wx_pay,
  validate_is_enabled_wx_pay,
};
use crate::r#gen::wx::wx_pay::wx_pay_model::WxPaySearch;

// wx_usr
use crate::r#gen::wx::wx_usr::wx_usr_dao::{
  find_by_id_wx_usr,
  validate_option_wx_usr,
};

// wxo_usr
use crate::r#gen::wx::wxo_usr::wxo_usr_dao::{
  find_by_id_wxo_usr,
  validate_option_wxo_usr,
};

use crate::r#gen::wx::pay_transactions_jsapi::pay_transactions_jsapi_model::{
  PayTransactionsJsapiTradeState,
  PayTransactionsJsapiInput,
};
use crate::r#gen::wx::pay_transactions_jsapi::pay_transactions_jsapi_dao::create_pay_transactions_jsapi;

use super::pay_transactions_jsapi_model::TransactionsJsapiInput;

use crate::common::oss::oss_dao::get_object;
use crate::r#gen::base::tenant::tenant_model::TenantId;

/// 生成商户订单号 out_trade_no
fn get_out_trade_no() -> String {
  get_short_uuid()
    .replace("+", "-")
    .replace("/", "_")
    .replace("=", "")
}

/// 微信支付 统一下单
#[allow(dead_code)]
pub async fn transactions_jsapi(
  transactions_jsapi_input: TransactionsJsapiInput,
  options: Option<Options>,
) -> Result<RequestPaymentOptions> {
  
  let appid = transactions_jsapi_input.appid;
  let description = transactions_jsapi_input.description;
  let amount = transactions_jsapi_input.amount;
  
  // 微信支付设置
  let wx_pay_model = validate_option_wx_pay(
    find_one_wx_pay(
      Some(WxPaySearch {
        appid: Some(appid),
        ..Default::default()
      }),
      None,
      options.clone(),
    ).await?
  ).await?;
  
  validate_is_enabled_wx_pay(
    &wx_pay_model,
  ).await?;
  
  // 当前登录用户有可能尚未绑定微信
  let auth_model = get_auth_model_err()?;
  
  let wx_usr_id = auth_model.wx_usr_id;
  let wxo_usr_id = auth_model.wxo_usr_id;
  
  let mut openid: Option<String> = None;
  let mut tenant_id: Option<TenantId> = None;
  
  if let Some(wx_usr_id) = wx_usr_id {
    let wx_usr_model = validate_option_wx_usr(
      find_by_id_wx_usr(
        wx_usr_id,
        options.clone(),
      ).await?
    ).await?;
    openid = Some(wx_usr_model.openid);
    tenant_id = Some(wx_usr_model.tenant_id);
  } else if let Some(wxo_usr_id) = wxo_usr_id {
    let wx_usr_model = validate_option_wxo_usr(
      find_by_id_wxo_usr(
        wxo_usr_id,
        options.clone(),
      ).await?
    ).await?;
    openid = Some(wx_usr_model.openid);
    tenant_id = Some(wx_usr_model.tenant_id);
  }
  
  if openid.is_none() || openid.as_ref().unwrap().is_empty() {
    return Err(eyre!("openid 不能为空"));
  }
  let openid = openid.unwrap();
  
  if tenant_id.is_none() || tenant_id.as_ref().unwrap().is_empty() {
    return Err(eyre!("tenant_id 不能为空"));
  }
  let tenant_id = tenant_id.unwrap();
  
  let private_key_str: Option<String> = {
    if wx_pay_model.private_key.is_empty() {
      None
    } else {
      let data = get_object(
        wx_pay_model.private_key.as_str(),
      ).await?;
      if let Some(data) = data {
        Some(String::from_utf8(data.into())?)
      } else {
        None
      }
    }
  };
  if private_key_str.is_none() || private_key_str.as_ref().unwrap().is_empty() {
    return Err(eyre!("私钥不存在"));
  }
  let private_key_str = private_key_str.unwrap();
  
  let appid = wx_pay_model.appid;
  let mchid = wx_pay_model.mchid;
  let serial_no = wx_pay_model.serial_no;
  let private_key = private_key_str;
  let v3_key = wx_pay_model.v3_key;
  let payer_client_ip = if wx_pay_model.payer_client_ip.is_empty() {
    None
  } else {
    Some(wx_pay_model.payer_client_ip)
  };
  let scene_info: Option<SceneInfo> = if let Some(payer_client_ip) = payer_client_ip {
    Some(SceneInfo {
      payer_client_ip: Some(payer_client_ip),
      ..Default::default()
    })
  } else {
    None
  };
  let notify_url: String = wx_pay_model.notify_url;
  
  if notify_url.is_empty() {
    return Err(eyre!("notify_url 不能为空"));
  }
  
  let wxpay = WxPay {
    appid: appid.as_str(),
    mchid: mchid.as_str(),
    private_key: private_key.as_str(),
    serial_no: serial_no.as_str(),
    api_v3_private_key: v3_key.as_str(),
    notify_url: notify_url.as_str(),
  };
  
  let out_trade_no = get_out_trade_no();
  
  let jsapi = Jsapi {
    description: description.clone(),
    out_trade_no: out_trade_no.clone(),
    amount: Amount {
      total: amount,
      ..Default::default()
    },
    payer: Payer {
      openid: openid.clone(),
    },
    time_expire: transactions_jsapi_input.time_expire.clone(),
    attach: transactions_jsapi_input.attach.clone(),
    scene_info,
    ..Default::default()
  };
  
  let wx_pay_data: WxPayData = wxpay.jsapi(&jsapi).await?;
  
  let request_payment_options = RequestPaymentOptions {
    time_stamp: wx_pay_data.time_stamp.to_string(),
    nonce_str: wx_pay_data.nonce_str,
    package: wx_pay_data.package.clone(),
    sign_type: wx_pay_data.sign_type,
    pay_sign: wx_pay_data.pay_sign,
  };
  
  let profit_sharing = if transactions_jsapi_input.profit_sharing.unwrap_or_default() {
    Some("Y".to_string())
  } else {
    Some("N".to_string())
  };
  
  if amount > u32::MAX as u64 {
    return Err(eyre!("金额超出范围"));
  }
  let amount = amount as u32;
  
  // 创建支付交易
  create_pay_transactions_jsapi(
    PayTransactionsJsapiInput {
      appid: Some(appid),
      mchid: Some(mchid),
      description: Some(description),
      out_trade_no: Some(out_trade_no),
      transaction_id: Some("".to_string()),
      trade_state: Some(PayTransactionsJsapiTradeState::Notpay),
      trade_state_desc: Some("未支付".to_string()),
      time_expire: transactions_jsapi_input.time_expire.clone(),
      attach: transactions_jsapi_input.attach.clone(),
      attach2: Some(transactions_jsapi_input.attach2),
      notify_url: Some(notify_url),
      receipt: transactions_jsapi_input.receipt,
      profit_sharing,
      total_fee: Some(amount),
      openid: Some(openid),
      prepay_id: Some(wx_pay_data.package),
      tenant_id: Some(tenant_id),
      ..Default::default()
    },
    options.clone(),
  ).await?;
  
  Ok(request_payment_options)
}